src/Aws/Kinesis/Client/Producer.hs

```diff
@@ -354,111 +318,101 @@ generatePartitionKey gen =
   Kin.partitionKey (T.pack name)
     & either (error ∘ T.unpack) id
 
--- | This will take up to @n@ items from a 'TBMQueue'.
+-- | This will take up to @n@ items from a 'TBMQueue', with a timeout.
 --
-takeTBMQueue
-  ∷ Int
+takeTBMQueueWithTimeout
+  ∷ ∀ α
+  . Int -- ^ number of items to get
+  → Int -- ^ timeout in microseconds
   → TBMQueue α
-  → STM [α]
-takeTBMQueue n q
-  | n <= 0 = return []
-  | otherwise = do
-      res ← tryReadTBMQueue q
-      case res of
-        Just (Just x) → (x:) <$> takeTBMQueue (n - 1) q
-        _ → return []
+  → IO [α]
+takeTBMQueueWithTimeout n timeout q = do
+  timedOutVar ← newEmptyTMVarIO
+
+  let
+    timeoutAction = do
+      threadDelay timeout
+      atomically $ putTMVar timedOutVar ()
+
+    go xs
+      | length xs ≥ n = return xs
+      | otherwise = do
+          res ← Left <$> readTBMQueue q <|> Right <$> readTMVar timedOutVar
+          case res of
+            Left mx → maybe (return xs) (go ∘ (:xs)) mx
+            Right () → return xs
+
+  withAsync timeoutAction $ \h → do
+    xs ← atomically $ go []
+    xs <$ cancel h
 
 -- | A policy for chunking the contents of the message queue.
 --
 data ChunkingPolicy
   = ChunkingPolicy
-  { _cpMaxChunkSize ∷ Int
+  { _cpMaxChunkSize ∷ !Int
   -- ^ The largest chunk size that is permitted.
 
-  , _cpThrottlingDelay ∷ Int
-  -- ^ The time after which a chunk should be committed, even if the maximum
-  -- chunk size has not yet been reached.
+  , _cpMinChunkingInterval ∷ !Int
+  -- ^ The time in microseconds after which a chunk should be committed, even
+  -- if the maximum chunk size has not yet been reached.
   }
 
 -- | A 'Source' that reads chunks off a bounded STM queue according some
 -- 'ChunkingPolicy'.
 --
 chunkedSourceTBMQueue
-  ∷ ( MonadIO m
-    , MonadBaseControl IO m
-    )
-  ⇒ ChunkingPolicy
+  ∷ ChunkingPolicy
   → TBMQueue α
-  → Source m [α]
-chunkedSourceTBMQueue bp@ChunkingPolicy{..} q = do
-  terminateNow ← liftIO ∘ atomically $ isClosedTBMQueue q
-  unless terminateNow $ do
-    items ← liftIO ∘ atomically $ takeTBMQueue _cpMaxChunkSize q
+  → Source IO [α]
+chunkedSourceTBMQueue cp@ChunkingPolicy{..} q = do
+  shouldTerminate ←
+    liftIO ∘ atomically $
+      (&&) <$> isClosedTBMQueue q <*> isEmptyTBMQueue q
+
+  unless shouldTerminate $ do
+    items ← lift $ takeTBMQueueWithTimeout _cpMaxChunkSize _cpMinChunkingInterval q
```

Comment:
I thought about it even more, and I think that the proposed semantics are precisely what we do not want for this function.

Imagine if there are five items in the queue, and we have requested ten. The whole point of the chunking mechanism here is to make sure that we keep yielding items at *at least* the specified interval/timeout; it may be that some chunks have fewer items than we requested, but this is desired. The purpose is to avoid a situation where certain messages never make it through to Kinesis, because we stopped writing to the queue before we hit the proper chunk-size.